src/tree/registries/UnifiedRegistryTreeDataProvider.ts

```diff
@@ -0,0 +1,196 @@
+
+        return {
+            dispose: () => {
+                this.providers.delete(provider.id);
+            }
+        };
+    }
+
+    public async refresh(): Promise<void> {
+        this.onDidChangeTreeDataEmitter.fire(undefined);
+    }
+
+    public async connectRegistryProvider(provider: RegistryDataProvider<unknown> | undefined = undefined): Promise<void> {
+        const connectedProviderIds = this.storageMemento.get<string[]>(ConnectedRegistryProvidersKey, []);
+
+        if (!provider) {
+            const picks: (vscode.QuickPickItem & { provider: RegistryDataProvider<unknown> })[] = [];
+
+            for (const currentProvider of this.providers.values()) {
+                if (connectedProviderIds.includes(currentProvider.id)) {
+                    continue;
+                }
+
+                picks.push({
+                    label: currentProvider.label,
+                    description: currentProvider.description,
+                    provider: currentProvider
+                });
+            }
+
+            const picked = await vscode.window.showQuickPick(picks, { placeHolder: vscode.l10n.t('Select a registry provider to use') });
+            if (!picked) {
+                return;
+            }
+
+            provider = picked.provider;
+        }
+
+        if (!connectedProviderIds.includes(provider.id)) {
+            await provider?.onConnect?.();
+            const connectedProviderIdsSet: Set<string> = new Set(connectedProviderIds);
+            connectedProviderIdsSet.add(provider.id);
+            await this.storageMemento.update(ConnectedRegistryProvidersKey, Array.from(connectedProviderIdsSet));
+        }
+
+        void this.refresh();
+    }
+
+    public async disconnectRegistryProvider(item: UnifiedRegistryItem<unknown>): Promise<void> {
+        await item.provider?.onDisconnect?.();
+
+        const newConnectedProviderIds = this.storageMemento
+            .get<string[]>(ConnectedRegistryProvidersKey, [])
+            .filter(cpi => cpi !== item.provider.id);
+        await this.storageMemento.update(ConnectedRegistryProvidersKey, newConnectedProviderIds);
+
+        void this.refresh();
+    }
+
+    /**
+     *
+     * @param imageBaseName The base name of the image to find registries for. e.g. 'docker.io'
+     * @returns A list of registries that are connected to the extension. If imageBaseName is provided, only registries that
+     *          can be used to push to that image will be returned.
+     */
+    public async getConnectedRegistries(imageBaseName?: string): Promise<UnifiedRegistryItem<CommonRegistry>[]> {
+        let registryRoots = await this.getChildren();
+        let findAzureRegistryOnly = false;
+
+        // filter out registry roots that don't match the image base name
+        if (imageBaseName) {
+            if (imageBaseName === 'docker.io') {
+                registryRoots = registryRoots.filter(r => (r.wrappedItem as CommonRegistryRoot).label === 'Docker Hub');

```

Comment:
For these three, we should use e.g. `ext.dockerHubRegistryDataProvider.label` instead of `'Docker Hub'`

Reply:
good point

Reply:
https://github.com/microsoft/vscode-docker/pull/4071 will address this